Add the Swan Lake Alpha2 Release Note (#1845)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-alpha2/RELEASE_NOTE.md

```ballerina
configurable table<AuthInfo> key(username) & readonly users = ?;
```

##### Support for Decrypting String Values Using the New Config Lang Library

The `bal encrypt` command can be used to encrypt plain text values and specify those in the `Config.toml` file. Then, the `config:decryptString()` function can be used to decrypt the configurable value. 

For example, if the `Config.toml` file contains the following encrypted string value,

```toml
password = "@encrypted:{ODYUoKSw0xW31eoxa/s2ESdBNgk1gX77txBIgpLC6NQ=}"
```

it can be decrypted in the Ballerina code as follows.

```ballerina
import ballerina/lang.config;

configurable string password = ?;

public function main() {
    string decryptedPassword = config:decryptString(password);
}
```

#### Standard Library

##### HTTP Module Improvements

###### Introduced Byte Stream Manipulation Functions to the Request and Response 

This introduction enables manipulating the payload as a stream of `byte[]`. The `setByteStream()` and `getByteStream()` methods use the Ballerina stream feature.

```ballerina
http:Request request = new;
io:ReadableByteChannel byteChannel = check io:openReadableFile("path/to/file.tmp");

> stream<io:Block, io:Error> byteStream = check byteChannel.blockStream(8196);request.setByteStream
> (byteStream) ;

Reviewer: ```suggestion
stream<io:Block, io:Error> byteStream = check byteChannel.blockStream(8196);
request.setByteStream(byteStream);
```

http:Response response = new;
stream<byte[], io:Error>|error str = response.getByteStream();
```

> ###### Introduced a Header Parameter to the Resource Signature

Reviewer: @chamil321, what about something like

```suggestion
###### Introduced the `http:Header` Annotation to Bind Headers in an Inbound Request
```

> With the introduction of the `@http:Header` annotation, inbound request headers can be retrieved by binding them to a resource signature parameter. Individual headers can be accessed as `string` or `string[]` types while a parameter of type `http:Headers` can be used to access all headers. 

Reviewer: Do we need "signature" here?
```suggestion
With the introduction of the `@http:Header` annotation, inbound request headers can be retrieved by binding them to a resource method parameter. Individual headers can be accessed as `string` or `string[]` types while a parameter of type `http:Headers` can be used to access all headers. 
```

```ballerina
service on helloEP {

>     resource function get hello(@http:Header {name: ”Accept”} string? acceptHeader, http:Headers allHeaders) {

Reviewer: ```suggestion
    resource function get hello(@http:Header {name: "Accept"} string? acceptHeader, http:Headers allHeaders) {
```

}

function getByteStream(int arraySize = 8196) returns stream<byte[], io:Error>|mime:ParserError {
}

function getBodyPartsAsStream(int arraySize = 8196) returns stream<byte[], io:Error>|mime:ParserError {

    // Sample
    byte[][] content = ["File Content".toBytes()];
    stream<byte[], io:Error> byteStream = content.toStream();
    mime:Entity entity = new;
    entity.setByteStream(byteStream);
    stream<byte[], io:Error>|mime:ParserError str = entity.getByteStream();
```

#### WebSocket Module Improvements

Introduced the Sync client. This is the primary client of the WebSocket module. This client is capable of reading and writing messages synchronously.

**Reading and Writing Text Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeTextMessage("Text message");
string textResp = check wsClient->readTextMessage();
```

**Reading and Writing Binary Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeBinaryMessage("Binary message".toBytes());
byte[] byteResp = check wsClient->readBinaryMessage();
```

> ###### GraphQL Module Improvements

Reviewer: ```suggestion
##### GraphQL Module Improvements
```

    websub:SubscriptionDeletedError? {
        return error websub:SubscriptionDeletedError("Subscriber wants to unsubscribe", headers = 
        {"Content-Encoding": "gzip"}, body = {"message": "Unsubscribing from the topic"});
    }
}
```

##### WebSubHub Module Improvements

Included functionality to the `websubhub:Service` to respond with user-defined custom payloads/header parameters in error scenarios.

```ballerina
import ballerina / websubhub;

service /websubhub on websubhub:Listener(9091) {

    remote function onRegisterTopic(websubhub:TopicRegistration message) returns websubhub:TopicRegistrationSuccess|
    websubhub:TopicRegistrationError {
        if (message.topic == "https://sub.topic.com") {
            websubhub:TopicRegistrationSuccess successResult = {body: <map<string>>{isSuccess: "true"}};
            return successResult;
        } else {
            return error websubhub:TopicRegistrationError("Topic registration failed!", headers = 
            {"Content-Encoding": "gzip"}, body = {"hub.additional.details": "Feature is not supported in the hub"});
        }
    }

> }

Reviewer 2: @MaryamZi here I have only provided one API method as a sample. There are 9 API methods to be implemented in total. Should we need to include all of them since this change is kind of common for all the APIs ?

> import ballerina / websubhub;
> 
> service /websubhub on websubhub:Listener(9091) {
> 
>     remote function onRegisterTopic(websubhub:TopicRegistration message) returns websubhub:TopicRegistrationSuccess|
>     websubhub:TopicRegistrationError {
>         if (message.topic == "https://sub.topic.com") {
>             websubhub:TopicRegistrationSuccess successResult = {body: <map<string>>{isSuccess: "true"}};
>             return successResult;
>         } else {
>             return error websubhub:TopicRegistrationError("Topic registration failed!", headers = 
>             {"Content-Encoding": "gzip"}, body = {"hub.additional.details": "Feature is not supported in the hub"});
>         }
>     }
> 
> }

Reviewer: Then shall we change this as 
```suggestion
import ballerina/websubhub;

service /websubhub on new websubhub:Listener(9091) {
    remote function onRegisterTopic(websubhub:TopicRegistration message) 
            returns websubhub:TopicRegistrationSuccess|websubhub:TopicRegistrationError {
        if (message.topic == "https://sub.topic.com") {
            websubhub:TopicRegistrationSuccess successResult = {body: <map<string>>{isSuccess: "true"}};
            return successResult;
        } else {
            return error websubhub:TopicRegistrationError("Topic registration failed!", 
                        headers = {"Content-Encoding": "gzip"}, 
                        body = {"hub.additional.details": "Feature is not supported in the hub"});
        }
    }

    // Other remote methods...
}
```

> }

Reviewer 2: Yes this is okay

```

##### IO Module Improvements

Introduce a parameter of type `XmlWriteOptions` to specify the entity type and the document type declaration.

```ballerina

> public function fileWriteXml(@untainted string path, xml content, *XmlWriteOptions xmlOptions, FileWriteOption fileWriteOption = 

Reviewer 3: Fixed in here https://github.com/ballerina-platform/ballerina-dev-website/pull/1845#discussion_r577387579

This `email:SecureSocket` record change would appear as follows.

```ballerina
public type SecureSocket record {|
    // ... other fields
    boolean verifyHostname = true;
|};
```

##### TCP Module Improvements

Introduced SSL/TLS-based communication to the TCP module using the `secureSocket` record configuration in both the client and the listener.

###### TCP Client with `secureSocket`

A sample code is as follows.

```ballerina
import ballerina/tcp;

configurable string keyPath = ?;
configurable string certPath = ?;

public function main() returns error? {
    tcp:Client socketClient = check new ("localhost", 9002, secureSocket = {
        certificate: {path: certPath},
        protocol: {
            name: "TLS",
            versions: ["TLSv1.2", "TLSv1.1"]
        },
        ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]
    });

    string msg = "Hello Ballerina Echo from secure client";
    byte[] msgByteArray = msg.toBytes();
    check socketClient->writeBytes(msgByteArray);

    readonly & byte[] receivedData = check socketClient->readBytes();
    test:assertEquals('string:fromBytes(receivedData), msg, "Found unexpected output");

    check socketClient->close();

> }

Reviewer 4: +1

```

###### TCP Listener with `secureSocket`

A sample code is as follows.

```ballerina
configurable string keyPath = ?;
configurable string certPath = ?;

service on new tcp:Listener(9002, secureSocket = {
    certificate: {path: certPath},
    privateKey: {path: keyPath},
    protocol: {
        name: "TLS",
        versions: ["TLSv1.2", "TLSv1.1"]
    },
    ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]
}) {

    isolated remote function onConnect(tcp:Caller caller) returns tcp:ConnectionService {
        io:println("Client connected to secureEchoServer: ", caller.remotePort);
        return new EchoService(caller);
    }
}

service class EchoService {
    tcp:Caller caller;

    public isolated function init(Caller c) {
        self.caller = c;
    }

    remote function onBytes(readonly & byte[] data) returns (readonly & byte[])|tcp:Error? {
        io:println("Echo: ", 'string:fromBytes(data));
        return data;
    }

> }

Reviewer 4: +1

```

##### UDP Module Improvements

> - The `sendDatagram` function will now send multiple datagrams if the size of the `byte[]` value provided as the `data` field of the datagram exceeds 8KB.

Reviewer: This is a method in a client, right? @MohamedSabthar 
```suggestion
- The `sendDatagram` method will now send multiple datagrams if the size of the `byte[]` value provided as the `data` field of the datagram exceeds 8KB.
```

Reviewer 4: +1

- Returning `Datagram` from the `onDatagram` or `onBytes` remote methods also sends multiple datagrams if the size of the `byte[]` value provided (as the `data` field of the datagram) exceeds 8KB.

##### Crypto Module Improvements

> Added support to decode private keys from `.key` files and public keys from `.cert` files and update the APIs for decoding private/public keys. This enables reading private/public keys from PEM files.

Reviewer: ```suggestion
Added support to decode private keys from `.key` files and public keys from `.cert` files and updated the APIs for decoding private/public keys. This enables reading private/public keys from PEM files.
```

Extended the private key support for JWT signature generation and public cert support for JWT signature validation.

#### Code to Cloud

- The `Kubernetes.toml` file is renamed to `Cloud.toml`. 
- The `--cloud=docker` build option is implemented. This will build the Dockerfile and Docker image without generating the Kubernetes artifacts.

#### Developer Tools

##### Language Server

Implemented renaming support in the language server. Now, VSCode users are able to select a symbol and rename all occurrences within the same module as well as across modules.

##### Debugger

Added variable paging support. With this feature, the Ballerina variables, which contain a large number of child variables will be shown in a paged view in the debug variable presentation.

##### Ballerina Shell REPL [EXPERIMENTAL]

1. Fixed the REPL expression output to output the `toBalString()` result.
2. Improved the REPL parser to support some partial snippets such as the example cases below.
	  - Template strings will allow continuing on new lines.
	  - The CLI will wait for more input if the last character was an operator.
	  - The CLI will not wait for unclosed double quotes.

> 3. Enable REPL to exit on `Ctrl+D`.

Reviewer: ```suggestion
3. Enabled REPL to exit on `Ctrl+D`.
```

#### Breaking Changes

> - Member access on a value of type `table` now returns `()` if the `table` does not contain a member with the specified key. Otherwise, the result is the member of the `table` with the given key.

Reviewer: ```suggestion
- Member access on a value of type `table` now returns `()` if the `table` does not contain a member with the specified key. Otherwise, the result is the member in the `table` value with the given key.
```

```ballerina

> type Employee record {
>     readonly string name;
>     int id;
> };
> 
> public function main() {
>     table<Employee> key(name) employeeTable = table [
>             {name: "Mike", id: 1234},
>             {name: "John", id: 4567}
>         ];
> 
>     Employee? emp1 = employeeTable["John"];
>     io:println(emp1); //{name: "John", id: 4567}
>     Employee? emp2 = employeeTable["Kate"];
>     boolean value2 = emp2 is (); //true
> }

Reviewer: ```suggestion
import ballerina/io;

type Employee record {
    readonly string name;
    int id;
};

public function main() {
    table<Employee> key(name) employeeTable = table [
        {name: "Mike", id: 1234},
        {name: "John", id: 4567}
    ];
    Employee? emp1 = employeeTable["John"];
    io:println(emp1); // {"name":"John","id":4567}
    Employee? emp2 = employeeTable["Kate"];
    io:println(emp2 is ()); // true
}
```

```

- Iterating over `xml` in a `from` clause in query expressions now returns `xml` and iterating over `xml<T>` returns `T`.

```ballerina

> xml authorList = xml `<authorList>
>                                       <author>
>                                           <name>Sir Arthur Conan Doyle</name>
>                                           <country>UK</country>
>                                       </author>
>                                      <author>
>                                         <name>Dan Brown</name>
>                                         <country>US</country>
>                                      </author>
>                                  </authorList>`;
> 
> xml authors = from xml y in authorList/<author>/<name>
>               select y;
> io:println
> (authors) ; //<name>Sir Arthur Conan Doyle</name><name>Dan Brown</name>

Reviewer: ```suggestion
import ballerina/io;

public function main() {
    xml authorList = xml `<authorList>
                            <author>
                                <name>Sir Arthur Conan Doyle</name>
                                <country>UK</country>
                            </author>
                            <author>
                                <name>Dan Brown</name>
                                <country>US</country>
                            </author>
                        </authorList>`;
    xml authors = from xml y in authorList/<author>/<name>
                  select y;
    io:println(authors); //<name>Sir Arthur Conan Doyle</name><name>Dan Brown</name>
}
```

```

> - The `readonly` and `value:Cloneable` values cannot be assigned to `any` since they contain an `error`.

Reviewer: ```suggestion
- Variables of types `readonly` and `value:Cloneable` cannot be assigned to `any` since they may contain values of type `error`.
```

Included functionality to the `websubhub:Service` to respond with user-defined custom payloads/header parameters in error scenarios.

```ballerina
import ballerina / websubhub;

service /websubhub on websubhub:Listener(9091) {

    remote function onRegisterTopic(websubhub:TopicRegistration message) returns websubhub:TopicRegistrationSuccess|
    websubhub:TopicRegistrationError {
        if (message.topic == "https://sub.topic.com") {
            websubhub:TopicRegistrationSuccess successResult = {body: <map<string>>{isSuccess: "true"}};
            return successResult;
        } else {
            return error websubhub:TopicRegistrationError("Topic registration failed!", headers = 
            {"Content-Encoding": "gzip"}, body = {"hub.additional.details": "Feature is not supported in the hub"});
        }
    }

}
```

##### IO Module Improvements

Introduce a parameter of type `XmlWriteOptions` to specify the entity type and the document type declaration.

```ballerina
public function fileWriteXml(@untainted string path, xml content, *XmlWriteOptions xmlOptions, FileWriteOption fileWriteOption = 
                             OVERWRITE) returns Error? {
}
```

> 

Reviewer 3: **Example**

```ballerina
import ballerina/io;

public function main() returns error? {
    xml content = xml `<note>
        <to>Tove</to>
        <from>Jani</from>
        <heading>Reminder</heading>
        <body>Don't forget me this weekend!</body>
        </note>`;
    string publicId = "-//W3C//DTD HTML 4.01 Transitional//EN";
    string systemId = "http://www.w3.org/TR/html4/loose.dtd";
    var writeResult = io:fileWriteXml("./xmlFileWithDoc.xml", content, doctype = {
        system: systemId,
        'public: publicId
    });
}
```

**Output**
```xml
<!DOCTYPE note PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```

Reviewer 3: @MaryamZi please check.

Reviewer: +1, can you please add it as a suggestion? I get a compilation error for this on the SNAPSHOT pack btw, hope this is now fixed.

> 

Reviewer 3: ```suggestion
**Example**

```ballerina
import ballerina/io;

public function main() returns error? {
    xml content = xml `<note>
        <to>Tove</to>
        <from>Jani</from>
        <heading>Reminder</heading>
        <body>Don't forget me this weekend!</body>
        </note>`;
    string publicId = "-//W3C//DTD HTML 4.01 Transitional//EN";
    string systemId = "http://www.w3.org/TR/html4/loose.dtd";
    var writeResult = io:fileWriteXml("./xmlFileWithDoc.xml", content, doctype = {
        system: systemId,
        'public: publicId
    });
}
```

**Output**
```xml
<!DOCTYPE note PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```
```

Reviewer 3: @MaryamZi @praneesha please approve.

> 

Reviewer 3: ````suggestion
**Example**

```ballerina
import ballerina/io;

public function main() returns error? {
    xml content = xml `<note>
        <to>Tove</to>
        <from>Jani</from>
        <heading>Reminder</heading>
        <body>Don't forget me this weekend!</body>
        </note>`;
    string publicId = "-//W3C//DTD HTML 4.01 Transitional//EN";
    string systemId = "http://www.w3.org/TR/html4/loose.dtd";
    var writeResult = io:fileWriteXml("./xmlFileWithDoc.xml", content, doctype = {
        system: systemId,
        'public: publicId
    });
}
```

**Output**
```xml
<!DOCTYPE note PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<note>
    <to>Tove</to>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend!</body>
</note>
```
````

Reviewer 3: @MaryamZi @praneesha please approve.

Reviewer: I can see an example in the note already? Do we need to change anything?

Reviewer 3: I addressed this comment https://github.com/ballerina-platform/ballerina-dev-website/pull/1845#discussion_r577323901

        - [WebSubHub Module Improvements](#websubhub-module-improvements)
        - [IO Module Improvements](#io-module-improvements)
        - [Email Module Improvements](#email-module-improvements)
        - [TCP Module Improvements](#tcp-module-improvements)
        - [UDP Module Improvements](#udp-module-improvements)
        - [Crypto Module Improvements](#crypto-module-improvements)
        - [JWT Module Improvements](#jwt-module-improvements)
    - [Code to Cloud](#code-to-cloud)
    - [Developer Tools](#developer-tools)
        - [Language Server](#language-server)
        - [Debugger](#debugger)
        - [Ballerina Shell REPL - EXPERIMENTAL](#ballerina-shell-repl-experimental)
    - [Breaking Changes](#breaking-changes)

### Updating Ballerina

You can use the [update tool](/learn/keeping-ballerina-up-to-date/) to update to Ballerina Swan Lake Alpha2 as follows.

#### For Existing Users

If you are already using Ballerina, you can directly update your distribution to the Swan Lake channel using the [Ballerina Update Tool](/learn/keeping-ballerina-up-to-date/). To do this, first, execute the command below to get the update tool updated to its latest version. 

> > `ballerina update`

Author: Fixed via 7d7c047ee0ccfd4bb8bb2e6fbc62400fd5a8334e.

From Swan Lake Alpha1 onwards, the `ballerina` command has to be issued as `bal`. Next, execute the command below to update to Swan Lake Alpha2.

> `bal dist pull slalpha2`

However, if you are using a Ballerina version below 1.1.0, install via the [installers](/downloads/#swanlake).

#### For New Users

If you have not installed Ballerina, then download the [installers](/downloads/#swanlake) to install.

### Highlights

- Support for mapping and error binding patterns in the `match` statement
- Support for configurable variables of `record` and `table` types

> - Support for decrypting string values using the new `config` lang library

Reviewer: ```suggestion
- Support for decrypting string values using the new `lang.config` lang library
```

- Improvements to the HTTP, mime, WebSocket, GraphQL, WebSub, WebSubHub, IO, email, TCP, UDP, crypto, and JWT standard library modules
- The extension of the Ballerina package distribution file has been changed from `.balo` to `.bala`

> - Improvements to developer tools such as the language server and debugger.

Reviewer: ```suggestion
- Improvements to developer tools such as the language server and the debugger
```

Author: No need to repeat "the" and no need for a full stop as these are bullet points and not complete sentences.

Reviewer: Yeah, need to remove the period at the end, right?
 ```suggestion
- Improvements to developer tools such as the language server and debugger
```

### What is new in Ballerina Swan Lake Alpha2

#### Language

##### Support for Mapping and Error Binding Patterns in the Match Statement

**Mapping Binding Pattern**

The `match` statement now supports mapping and error binding patterns with `var`.

```ballerina

> match v {
> var  {a, b} => {
>         // Matches mappings that contain at least fields `a` and `b`.
>         // The values of these fields can be accessed via the variables 
>         // `a` and `b` within this block.
>         io: println(a);
> }
> 
> var {c:{x:a1, y:a2}, ...d} => {
> // Matches mappings that have a field `c` where its value is 
> // another mapping that contains at least the fields `x` and `y`.
> // All of the remaining fields (if any) can be accessed via
> // the new variable `d`.
> int length = d.length();
> }
> }

Reviewer: ```suggestion
match v {
    var  {a, b} => {
        // Matches mappings that contain at least the fields `a` and `b`.
        // The values of these fields can be accessed via the variables 
        // `a` and `b` within this block.
        io:println(a);
    }

    var {c: {x: a1, y: a2}, ...rest} => {
        // Matches mappings that have a field `c` where its value is 
        // another mapping that contains at least the fields `x` and `y`.
        // All of the remaining fields (if any) can be accessed via
        // the new variable `rest`.
        int length = rest.length();
    }
}
```

```

**Error Binding Pattern**

```ballerina

> match v {
> var  error(message, error(causeMessage)) => {
>         // Matches errors that have a cause. 
>         // The messages of the matched error and the cause error
>         // can be accessed via the variables `message` and 
>         // `causeMessage` within this block.
>         io: println(causeMessage);
> }
> 
> var error(a, code = matchedCode) => {
>     // Matches errors that have a detail entry with the key `code`.
>     // The `code` can be accessed using the `matchedCode` variable 
>     // within this block.
>     io: println(matchedCode);
> }
> }

Reviewer: ```suggestion
match v {
    var error(message, error(causeMessage)) => {
        // Matches errors that have a cause. 
        // The messages of the matched error and the cause error
        // can be accessed via the variables `message` and 
        // `causeMessage` within this block.
        io:println(causeMessage);
    }

    var error(a, code = matchedCode) => {
        // Matches errors that have a detail entry with the key `code`.
        // The `code` can be accessed using the `matchedCode` variable 
        // within this block.
        io:println(matchedCode);
    }
}
```

```

#### Runtime

##### Support for Configurable Variables of Record and Table Types

**Record Type**

Record fields with simple types like `int`, `string`, `boolean`, `float`, `decimal`, and arrays of the respective types are now supported. 

For example, if the `Config.toml` file contains the following TOML table,

```toml
[Pkg.testUserOne]
username = "john"
password = "abcd"
scopes = ["write"]

[Pkg.testUserTwo]

> Username = “mary”
> Password = “test123”

Reviewer 5: ```suggestion
username = "mary"
password = "test123"
```

> Username = “mary”
> Password = “test123”

Reviewer 6: Shouldn't these should be updated as follows? @warunalakshitha @HindujaB 
```suggestion
username = “mary”
password = “test123”
```

> username = “mary”
> password = “test123”

Reviewer: ```suggestion
username = "mary"
password = "test123"
```

```

> it can be loaded as a configurable variable of a record type as follows.

Reviewer: ```suggestion
it can be loaded as configurable variables of a record type as follows.
```

```ballerina
type AuthInfo record {
    readonly string username;
    string password;
    string[] scopes?;
};

configurable AuthInfo & readonly testUserOne = ?;
configurable AuthInfo & readonly testUserTwo = ?;
```

**Table Type**

Ballerina now supports configurable variables of type `table` through TOML arrays of tables.

For example, if the `Config.toml` file contains the following TOML table array,

```toml
[[Pkg.users]]
username = "alice"

> password="password1"
> scopes=["scope1"]

Reviewer: ```suggestion
password = "password1"
scopes = ["scope1"]
```

[[Pkg.users]]
username = "bob"

> password="password2"
> scopes=["scope1", "scope2"]

Reviewer: ```suggestion
password = "password2"
scopes = ["scope1", "scope2"]
```

[[Pkg.users]]
username = "jack"

> password="password3"

Reviewer: ```suggestion
password = "password3"
```

```

it can be loaded as a configurable variable of a `table` type as follows.

```ballerina
configurable table<AuthInfo> key(username) & readonly users = ?;
```

##### Support for Decrypting String Values Using the New Config Lang Library

> The `bal encrypt` command can be used to encrypt the plain text values and specify those in the `Config.toml` file. Then, the `config:decryptString(<string variable>)` can be used to decrypt the configurable value. 

Reviewer: ```suggestion
The `bal encrypt` command can be used to encrypt plain text values and specify those in the `Config.toml` file. Then, the `config:decryptString()` function can be used to decrypt the configurable value. 
```

For example, if the `Config.toml` file contains the following encrypted string value,

```toml
password = "@encrypted:{ODYUoKSw0xW31eoxa/s2ESdBNgk1gX77txBIgpLC6NQ=}"
```

> it will be decrypted in the Ballerina code as follows.

Reviewer: ```suggestion
it can be decrypted in the Ballerina code as follows.
```

    entity.setByteStream(byteStream);
    stream<byte[], io:Error>|mime:ParserError str = entity.getByteStream();
```

#### WebSocket Module Improvements

Introduced the Sync client. This is the primary client of the WebSocket module. This client is capable of reading and writing messages synchronously.

**Reading and Writing Text Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeTextMessage("Text message");
string textResp = check wsClient->readTextMessage();
```

**Reading and Writing Binary Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeBinaryMessage("Binary message".toBytes());
byte[] byteResp = check wsClient->readBinaryMessage();
```

###### GraphQL Module Improvements

Ballerina GraphQL listeners can now be configured using the same configurations as the listener configurations in the `http:Listener`. Additionally, a GraphQL service can be secured by defining a `maxQueryDepth` as an annotation to restrict the depth of a query before execution.

```ballerina
import ballerina/graphql;

graphql:ListenerConfiguration configs = {

> // http listener configurations

Reviewer 7: ```suggestion
    // http listener configurations
```

};
listener graphql:Listener graphqlListener = new (9090, configs);

@graphql:ServiceConfigurration {maxQueryDepth: 3}
 Service /graphql on graphqlListener {

> // Service definition

Reviewer 7: ```suggestion
    // Service definition
```

            websubhub:TopicRegistrationSuccess successResult = {body: <map<string>>{isSuccess: "true"}};
            return successResult;
        } else {
            return error websubhub:TopicRegistrationError("Topic registration failed!", headers = 
            {"Content-Encoding": "gzip"}, body = {"hub.additional.details": "Feature is not supported in the hub"});
        }
    }

}
```

##### IO Module Improvements

Introduce a parameter of type `XmlWriteOptions` to specify the entity type and the document type declaration.

```ballerina
public function fileWriteXml(@untainted string path, xml content, *XmlWriteOptions xmlOptions, FileWriteOption fileWriteOption = 
                             OVERWRITE) returns Error? {
}
```

##### Email Module Improvements

- Make the `body` field of the `email:Message` record optional. This enables sending an email only with the `htmlBody` without a text-typed body field.

- Add `mime:Entity` type to the union type of the `attachments` field in the `email:Message` record. That enables attaching a single MIME Entity as an attachment. This `email:Message` record change would appear as follows.

```ballerina
public type Message record {|

>     // … other fields

Reviewer 8: ```suggestion
    // ... other fields
```

    mime:Entity|Attachment|(mime:Entity|Attachment)[] attachments?;
|};
```

- The above-mentioned changes related to the `body` and `attachments` fields are updated in the `email:Options` record as given below in order to facilitate the `sendEmail` method in the `email:SmtpClient`.

```ballerina
public type Options record {|

>     // … other fields

Reviewer 8: ```suggestion
    // ... other fields
```

    string body?;
    mime:Entity|Attachment|(mime:Entity|Attachment)[] attachments?;
|};
```

- Remove the `properties` field representing custom properties from all Email module related configuration records: `email:SmtpConfig`, `email:PopConfig`, `email:ImapConfig`, `email:PopListenerConfig`, and `email:ImapListenerConfig`.

- The `mail.smtp.ssl.checkserveridentity` custom property was passed as an entry in the `properties` to enable/disable the server certificate’s hostname verification. As the `properties` field is removed from the API, from this release onwards, the `verifyHostName` boolean field is introduced to the `secureSocket` record for all the configurations related to the Email module. 

This `email:SecureSocket` record change would appear as follows.

```ballerina
public type SecureSocket record {|

> // … other fields

Reviewer 8: ```suggestion
// ... other fields
```

> public type SecureSocket record {|
> // ... other fields
> boolean verifyHostname = true;
> |};

Reviewer 6: ```suggestion
public type SecureSocket record {|
    // ... other fields
    boolean verifyHostname = true;
|};
```

service on new tcp:Listener(9002, secureSocket = {
    certificate: {path: certPath},
    privateKey: {path: keyPath},
    protocol: {
        name: "TLS",
        versions: ["TLSv1.2", "TLSv1.1"]
    },
    ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]
}) {

    isolated remote function onConnect(tcp:Caller caller) returns tcp:ConnectionService {
        io:println("Client connected to secureEchoServer: ", caller.remotePort);
        return new EchoService(caller);
    }
}

service class EchoService {
    tcp:Caller caller;

    public isolated function init(Caller c) {
        self.caller = c;
    }

    remote function onBytes(readonly & byte[] data) returns (readonly & byte[])|tcp:Error? {
        io:println("Echo: ", 'string:fromBytes(data));
        return data;
    }
}
```

##### UDP Module Improvements

- The `sendDatagram` function will now send multiple datagrams if the size of the `byte[]` value provided as the `data` field of the datagram exceeds 8KB.

- Returning `Datagram` from the `onDatagram` or `onBytes` remote methods also sends multiple datagrams if the size of the `byte[]` value provided (as the `data` field of the datagram) exceeds 8KB.

> ###### Crypto Module Improvements

Reviewer 6: ```suggestion
##### Crypto Module Improvements
```

Added support to decode private keys from `.key` files and public keys from `.cert` files and update the APIs for decoding private/public keys. This enables reading private/public keys from PEM files.

> ###### JWT Module Improvements

Reviewer 6: ```suggestion
##### JWT Module Improvements
```

Extended the private key support for JWT signature generation and public cert support for JWT signature validation.

#### Code to Cloud

- The `Kubernetes.toml` file is renamed to `Cloud.toml`. 
- The `--cloud=docker` build option is implemented. This will build the Dockerfile and Docker image without generating the Kubernetes artifacts.

#### Developer Tools

##### Language Server

> Implemented renaming any symbol in the Language Server.

Reviewer 9: Implemented renaming support in the language server. VS Code users are now able to select a symbol and rename all occurrences within the same module as well as across modules.

Author: @nadeeshaan

Author: Will be fixed via https://github.com/ballerina-platform/ballerina-dev-website/pull/1845#discussion_r576914689.

> Implemented renaming any symbol in the Language Server.

Reviewer 9: ```suggestion
Implemented renaming support in the language server. VS Code users are now able to select a symbol and rename all occurrences within the same module as well as across modules.
```

##### Debugger

Added variable paging support. With this feature, the Ballerina variables, which contain a large number of child variables will be shown in a paged view in the debug variable presentation.

##### Ballerina Shell REPL [EXPERIMENTAL]

> 1. Fixed the REPL expression output to output the `toBalString()` result.
> 2. Improved the REPL parser to support some partial snippets such as the example cases below.
> 
> - Template strings will allow continuing on new lines.
> - The CLI will wait for more input if the last character was an operator.
> - The CLI will not wait for unclosed double quotes.
> 
> 3. Enable REPL to exit on `Ctrl+D`.

Reviewer 10: ```suggestion
1. Fixed the REPL expression output to output the `toBalString()` result.
2. Improved the REPL parser to support some partial snippets such as the example cases below.
	  - Template strings will allow continuing on new lines.
	  - The CLI will wait for more input if the last character was an operator.
	  - The CLI will not wait for unclosed double quotes.
3. Enable REPL to exit on `Ctrl+D`.
```

```

#### WebSocket Module Improvements

Introduced the Sync client. This is the primary client of the WebSocket module. This client is capable of reading and writing messages synchronously.

**Reading and Writing Text Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeTextMessage("Text message");
string textResp = check wsClient->readTextMessage();
```

**Reading and Writing Binary Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeBinaryMessage("Binary message".toBytes());
byte[] byteResp = check wsClient->readBinaryMessage();
```

##### GraphQL Module Improvements

Ballerina GraphQL listeners can now be configured using the same configurations as the listener configurations in `http:Listener`. Additionally, a GraphQL service can be secured by defining a `maxQueryDepth` as an annotation to restrict the depth of a query before execution.

```ballerina
import ballerina/graphql;

graphql:ListenerConfiguration configs = {
    // http listener configurations
};
listener graphql:Listener graphqlListener = new (9090, configs);

> @graphql:ServiceConfigurration {maxQueryDepth: 3}
>  service /graphql on graphqlListener {

Reviewer 7: ```suggestion
@graphql:ServiceConfiguration {maxQueryDepth: 3}
service /graphql on graphqlListener {
```

    }

    remote function onBytes(readonly & byte[] data) returns (readonly & byte[])|tcp:Error? {
        io:println("Echo: ", 'string:fromBytes(data));
        return data;
    }
}
```

##### UDP Module Improvements

- The `sendDatagram` function will now send multiple datagrams if the size of the `byte[]` value provided as the `data` field of the datagram exceeds 8KB.

- Returning `Datagram` from the `onDatagram` or `onBytes` remote methods also sends multiple datagrams if the size of the `byte[]` value provided (as the `data` field of the datagram) exceeds 8KB.

##### Crypto Module Improvements

Added support to decode private keys from `.key` files and public keys from `.cert` files and updated the APIs for decoding private/public keys. This enables reading private/public keys from PEM files.

##### JWT Module Improvements

Extended the private key support for JWT signature generation and public cert support for JWT signature validation.

##### Log module improvements

Introduced a configuration in the log module to set the output format to JSON. You need to add the entry below in the `Config.toml` file to set the output format to JSON.

```toml
[log]

> output_format = "json"

Reviewer 11: ```suggestion
format = "json"
```

        - [IO Module Improvements](#io-module-improvements)
        - [Email Module Improvements](#email-module-improvements)
        - [TCP Module Improvements](#tcp-module-improvements)
        - [UDP Module Improvements](#udp-module-improvements)
        - [Crypto Module Improvements](#crypto-module-improvements)
        - [JWT Module Improvements](#jwt-module-improvements)
    - [Code to Cloud](#code-to-cloud)
    - [Developer Tools](#developer-tools)
        - [Language Server](#language-server)
        - [Debugger](#debugger)
        - [Ballerina Shell REPL - EXPERIMENTAL](#ballerina-shell-repl-experimental)
    - [Breaking Changes](#breaking-changes)

### Updating Ballerina

You can use the [update tool](/learn/keeping-ballerina-up-to-date/) to update to Ballerina Swan Lake Alpha2 as follows.

#### For Existing Users

If you are already using Ballerina, you can directly update your distribution to the Swan Lake channel using the [Ballerina Update Tool](/learn/keeping-ballerina-up-to-date/). To do this, first, execute the command below to get the update tool updated to its latest version. 

> `bal update`

> If you are using an **Update Tool version below 0.8.14**, execute the `ballerina update` command to update it. Next, execute the command below to update to Swan Lake Alpha2.

Reviewer 12: Same concern as above regarding this

Reviewer 12: When a user is reading this, might try out the command at the top first, only to find out it doesn't work and then, a few lines down only it mentions the compatible version info

Author: @keizer619

Reviewer 13: Lets include another section for ballerina update for users who using older version than 0.8.14 as in download page

Reviewer 13: https://github.com/ballerina-platform/ballerina-dev-website/pull/1845/files#r578960154

Author: @pubudu91

> `bal dist pull slalpha2`

#### For New Users

If you have not installed Ballerina, then download the [installers](/downloads/#swanlake) to install.

### Highlights

> - Support for mapping and error binding patterns in the `match` statement

Reviewer 12: Shouldn't this be "mapping and error match patterns"? @hasithaa @MaryamZi

Reviewer 12: Ah nvm. Just saw the relevant section. Misunderstood what this says.

Author: @pubudu91 - Can we resolve this if this seems fine?

- Support for configurable variables of `record` and `table` types
- Support for decrypting string values using the new `lang.config` lang library

> - Improvements to the HTTP, mime, WebSocket, GraphQL, WebSub, WebSubHub, IO, email, TCP, UDP, crypto, and JWT standard library modules

Reviewer 12: mime -> MIME right? @chamil321

Reviewer 12: Otherwise looks like the odd one out since others use their properly-cased terms

Reviewer 14: Yes, It should be MIME which is the short form of `Multipurpose Internet Mail Extension`

> - Improvements to the HTTP, mime, WebSocket, GraphQL, WebSub, WebSubHub, IO, email, TCP, UDP, crypto, and JWT standard library modules

Reviewer 14: ```suggestion
- Improvements to the HTTP, MIME, WebSocket, GraphQL, WebSub, WebSubHub, IO, email, TCP, UDP, crypto, and JWT standard library modules
```

> - The extension of the Ballerina package distribution file has been changed from `.balo` to `.bala`

Reviewer 12: Can't we simply refer to this as a "Ballerina package"? @azinneera

Reviewer 15: The package we create now is not a binary as it used to be some time ago. This is an archive.

Author: @pubudu91

> - Improvements to developer tools such as the language server and debugger

Reviewer 12: ```suggestion
- Improvements to developer tools such as the Language Server and debugger
```

### What is new in Ballerina Swan Lake Alpha2

#### Language

> ##### Support for Mapping and Error Binding Patterns in the Match Statement

Reviewer 12: Binding Patterns -> Match Patterns I think

##### HTTP Module Improvements

###### Introduced Byte Stream Manipulation Functions to the Request and Response 

This introduction enables manipulating the payload as a stream of `byte[]`. The `setByteStream()` and `getByteStream()` methods use the Ballerina stream feature.

```ballerina
http:Request request = new;
io:ReadableByteChannel byteChannel = check io:openReadableFile("path/to/file.tmp");
stream<io:Block, io:Error> byteStream = check byteChannel.blockStream(8196);
request.setByteStream(byteStream);

http:Response response = new;
stream<byte[], io:Error>|error str = response.getByteStream();
```

###### Introduced the `http:Header` Annotation to Bind Headers in an Inbound Request

With the introduction of the `@http:Header` annotation, inbound request headers can be retrieved by binding them to a resource method parameter. Individual headers can be accessed as `string` or `string[]` types while a parameter of type `http:Headers` can be used to access all headers. 

```ballerina
service on helloEP {
    resource function get hello(@http:Header {name: "Accept"} string? acceptHeader, http:Headers allHeaders) {
    //...
    }
}
```

##### MIME Module Improvements

> ###### Introduced Byte Stream Manipulation Functions to the `mime:Entity` Class

Reviewer 12: ```suggestion
###### Introduced Byte Stream Manipulation Methods to the `mime:Entity` Class
```

This introduction enables manipulating the entity body as a stream of `byte[]`.

```ballerina

> function setByteStream(stream<byte[], io:Error> byteStream, string contentType = "application/octet-stream") {

Reviewer 14: ```suggestion
function setByteStream(stream<byte[], io:Error> byteStream, string contentType = "application/octet-stream")
```

> }

Reviewer 14: ```suggestion
```

> function getByteStream(int arraySize = 8196) returns stream<byte[], io:Error>|mime:ParserError {

Reviewer 14: ```suggestion
function getByteStream(int arraySize = 8196) returns stream<byte[], io:Error>|mime:ParserError
```

> }

Reviewer 14: ```suggestion
```

> }

Reviewer 12: What's the purpose of these two empty functions? If the idea was to show the signature of the two methods mentioned, let's put it as a method signature: e.g.,
```ballerina
function setByteStream(stream<byte[], io:Error> byteStream, string contentType = "application/octet-stream")
```
Or maybe link to the API docs?

Author: @chamil321

> function getBodyPartsAsStream(int arraySize = 8196) returns stream<byte[], io:Error>|mime:ParserError {

Reviewer 14: ```suggestion
function getBodyPartsAsStream(int arraySize = 8196) returns stream<byte[], io:Error>|mime:ParserError
```

>     // Sample
>     byte[][] content = ["File Content".toBytes()];
>     stream<byte[], io:Error> byteStream = content.toStream();
>     mime:Entity entity = new;
>     entity.setByteStream(byteStream);
>     stream<byte[], io:Error>|mime:ParserError str = entity.getByteStream();
> }

Reviewer 14: ```suggestion
```

Introduced the Sync client. This is the primary client of the WebSocket module. This client is capable of reading and writing messages synchronously.

**Reading and Writing Text Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeTextMessage("Text message");
string textResp = check wsClient->readTextMessage();
```

**Reading and Writing Binary Messages**

```ballerina
websocket:Client wsClient = check new ("ws://echo.websocket.org");
var err = wsClient->writeBinaryMessage("Binary message".toBytes());
byte[] byteResp = check wsClient->readBinaryMessage();
```

##### GraphQL Module Improvements

Ballerina GraphQL listeners can now be configured using the same configurations as the listener configurations in `http:Listener`. Additionally, a GraphQL service can be secured by defining a `maxQueryDepth` as an annotation to restrict the depth of a query before execution.

```ballerina
import ballerina/graphql;

graphql:ListenerConfiguration configs = {
    // http listener configurations
};
listener graphql:Listener graphqlListener = new (9090, configs);

@graphql:ServiceConfigurration {maxQueryDepth: 3}

>  Service /graphql on graphqlListener {

Reviewer 12: ```suggestion
 service /graphql on graphqlListener {
```

    // Service definition
}
```

##### WebSub Module Improvements

Included functionality to the `websub:SubscriberService` to respond with user-defined custom payloads/header parameters in error scenarios.

```ballerina
import ballerina/websub;

listener websub:Listener subscriberListener = new (9001);

service /subscriber on subscriberListener {
    remote function onSubscriptionValidationDenied(websub:SubscriptionDeniedError msg) returns websub:Acknowledgement? {
        websub:Acknowledgement ack = {
            headers: {
                "Content-Encoding": "gzip"
            },
            body: {
                "message": "Successfully processed request"
            }
        };
        return ack;
    }

    remote function onSubscriptionVerification(websub:SubscriptionVerification msg) returns 
            websub:SubscriptionVerificationSuccess|websub:SubscriptionVerificationError {
        if (msg.hubTopic == "https://www.sample.topic") {
            return error websub:SubscriptionVerificationError("Hub topic not supported", 
                                                              headers = {"Content-Encoding": "gzip"}, 

>                                                               body = {"message": "Hub topic not supported"});

Reviewer 12: Shouldn't this be the else condition's body? @shafreenAnfar

Author: @Bhashinee

Reviewer 16: Tagging @ayeshLK

Reviewer 2: @pubudu91 in `WebSub Hub` either it could block / allow subscription based on `topic`. Here the `Hub` is denying the subscription verification based on the given `topic`.

Reviewer 12: Ack.

```ballerina
public type Options record {|
    // ... other fields
    string body?;
    mime:Entity|Attachment|(mime:Entity|Attachment)[] attachments?;
|};
```

- Remove the `properties` field representing custom properties from all Email module related configuration records: `email:SmtpConfig`, `email:PopConfig`, `email:ImapConfig`, `email:PopListenerConfig`, and `email:ImapListenerConfig`.

- The `mail.smtp.ssl.checkserveridentity` custom property was passed as an entry in the `properties` to enable/disable the server certificate’s hostname verification. As the `properties` field is removed from the API, from this release onwards, the `verifyHostName` boolean field is introduced to the `secureSocket` record for all the configurations related to the Email module. 

This `email:SecureSocket` record change would appear as follows.

```ballerina
public type SecureSocket record {|
    // ... other fields
    boolean verifyHostname = true;
|};
```

##### TCP Module Improvements

Introduced SSL/TLS-based communication to the TCP module using the `secureSocket` record configuration in both the client and the listener.

> ###### TCP Client with `secureSocket`

Reviewer 12: I think this needs to be reworded. `secureSocket` is the param name right? Wouldn't "Secure TCP Client" or "Support for Securing TCP Client" be more appropriate? @shafreenAnfar @chamil321

Author: @Maninda

Reviewer 14: @MohamedSabthar

Reviewer 4: Need to remove the `tcp module improvement` section

Author: Removed.

Sample code is as follows.

```ballerina
import ballerina/io;
import ballerina/tcp;

configurable string certPath = ?;

public function main() returns error? {
    tcp:Client socketClient = check new ("localhost", 9002, secureSocket = {
        certificate: {path: certPath},
        protocol: {
            name: "TLS",
            versions: ["TLSv1.2", "TLSv1.1"]
        },
        ciphers: ["TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"]
    });
    string msg = "Hello Ballerina Echo from secure client";
    byte[] msgByteArray = msg.toBytes();
    check socketClient->writeBytes(msgByteArray);
    readonly & byte[] receivedData = check socketClient->readBytes();
    io:println(string:fromBytes(receivedData));
    check socketClient->close();
}
```

> ###### TCP Listener with `secureSocket`

Reviewer 12: Same comment as above regarding the use of `secureSocket` in the heading

Author: @chamil321

Reviewer 14: @ldclakmal

Reviewer 6: I think it is decided to remove this section. @MohamedSabthar

Reviewer 4: Need to remove the `tcp module improvement` section since it only contains the secure socket implementation

Author: Removed.

        return new EchoService(caller);
    }
}
service class EchoService {
    tcp:Caller caller;

    public isolated function init(tcp:Caller c) {
        self.caller = c;
    }

    remote function onBytes(readonly & byte[] data) returns (readonly & byte[])|tcp:Error? {
        io:println("Echo: ", string:fromBytes(data));
        return data;
    }
}
```

##### UDP Module Improvements

- The `sendDatagram` method will now send multiple datagrams if the size of the `byte[]` value provided as the `data` field of the datagram exceeds 8KB.

- Returning `Datagram` from the `onDatagram` or `onBytes` remote methods also sends multiple datagrams if the size of the `byte[]` value provided (as the `data` field of the datagram) exceeds 8KB.

##### Crypto Module Improvements

Added support to decode private keys from `.key` files and public keys from `.cert` files and updated the APIs for decoding private/public keys. This enables reading private/public keys from PEM files.

##### JWT Module Improvements

Extended the private key support for JWT signature generation and public cert support for JWT signature validation.

> ##### Log module improvements

Reviewer 12: ```suggestion
##### Log Module Improvements
```

Introduced a configuration in the log module to set the output format to JSON. You need to add the entry below in the `Config.toml` file to set the output format to JSON.

```toml
[log]
format = "json"
```

#### Code to Cloud

- The `Kubernetes.toml` file is renamed to `Cloud.toml`. 
- The `--cloud=docker` build option is implemented. This will build the Dockerfile and Docker image without generating the Kubernetes artifacts.

#### Developer Tools

##### Language Server

> Implemented renaming support in the language server. Now, VSCode users are able to select a symbol and rename all occurrences within the same module as well as across modules.

Reviewer 12: ```suggestion
Implemented renaming support in the Language Server. Now VS Code users are able to select a symbol and rename all occurrences within the same module as well as across modules.
```

Author: @pubudu91 - We need the comma after "Now".

Author: Fixed.

##### Debugger

Added variable paging support. With this feature, the Ballerina variables, which contain a large number of child variables will be shown in a paged view in the debug variable presentation.

##### Ballerina Shell REPL [EXPERIMENTAL]

1. Fixed the REPL expression output to output the `toBalString()` result.
2. Improved the REPL parser to support some partial snippets such as the example cases below.
	  - Template strings will allow continuing on new lines.
	  - The CLI will wait for more input if the last character was an operator.
	  - The CLI will not wait for unclosed double quotes.

> 3. Enabled REPL to exit on `Ctrl+D`.

Reviewer 12: Shall we make this section also bullets? Otherwise looks inconsistent with the rest of the sections right?

Author: Done.
